common/scpi.c

```diff
@@ -0,0 +1,217 @@
+			buffer->busy = false;
+			return;
+		}
+	}
+
+	/* Copy the prepared reply from the buffer to shared memory. */
+	memcpy(&SCPI_MEM_AREA(client).tx_msg, &buffer->mem.tx_msg,
+	       SCPI_HEADER_SIZE + buffer->mem.tx_msg.size);
+
+	/* Notify the client that the message has been sent. */
+	if ((err = msgbox_send_msg(scpi_msgbox, client, SCPI_VIRTUAL_CHANNEL)))
+		error("SCPI.%u: Error sending reply: %d", client, err);
+
+	/* Mark the buffer as no longer in use. */
+	buffer->busy = false;
+}
+
+/**
+ * Parts 1 & 2 of SCPI message handling (SCP-initiated messages).
+ */
+int
+scpi_create_message(uint8_t client, uint8_t command, uint32_t *payload,
+                    uint16_t payload_size)
+{
+	struct scpi_buffer *buffer = scpi_alloc_buffer(client);
+
+	/* Ensure there is space to put the created message. */
+	if (buffer == NULL)
+		return EAGAIN;
+
+	/* Create the message header. */
+	buffer->mem.tx_msg.command = command;
+	buffer->mem.tx_msg.sender  = 0;
+	buffer->mem.tx_msg.size    = payload_size;
+	buffer->mem.tx_msg.status  = SCPI_OK;
+
+	/* Copy the payload to the buffer, if there is one. */
+	if (payload_size > 0)
+		memcpy(&buffer->mem.tx_msg, payload, payload_size);
+
+	queue_work(scpi_send_message, buffer);
+
+	return SUCCESS;
+}
+
+/**
+ * Part 2 of SCPI message handling.
+ *
+ * Dispatch handling of the message to the appropriate function for each
+ * command. The functions for doing so are defined in a separate file to
+ * separate the API functionality from communication/synchronization code.
+ *
+ * When the command's actions are complete, and a reply message is created,
+ * enqueue a function to send the response.
+ */
+static void
+scpi_handle_message(void *param)
+{
+	struct scpi_buffer *buffer = param;
+
+	assert(buffer->busy);
+
+	/* Create a reply in tx_msg based on the contents of rx_msg. */
+	if (!scpi_handle_cmd(buffer->client, &buffer->mem.rx_msg,
+	                     &buffer->mem.tx_msg)) {
+		/* If the message does not need a reply, stop processing. */
+		buffer->busy = false;
+		return;
+	}
+
+	queue_work(scpi_send_message, buffer);
+}
+
+/**
+ * Part 1 of SCPI message handling (received messages).
+ *
+ * This is the callback from the message box framework; it is called when a new
+ * message is received. The SCPI client regains ownership of the shared memory
+ * area once the message is acknowledged, which happens when the message box
+ * driver clears the pending IRQ, immediately after this function returns.
+ * Thus, we must do our best to copy the shared memory area's contents to a
+ * safe before returning.
+ *
+ * Once the message is in a location controlled by the SCP firmware, enqueue a
+ * work item to parse, handle, and respond to it. This allows the interrupt to
+ * finish as quickly as possible.
+ */
+static void
+scpi_receive_message(struct device *dev __unused, uint8_t client, uint32_t msg)
+{
+	struct scpi_buffer *buffer;
+	struct scpi_msg    *rx_msg = &SCPI_MEM_AREA(client).rx_msg;
```

Comment:
Alignment is off... did you `make format`?

Reply:
I did (see the CI). The variable names are aligned to each other. (Maybe you were looking at an older commit, before I fixed alignment issues)

Reply:
It appears you force pushed while I was reviewing (see out-of-date changes above). Thanks for correcting.